### DISCUSSION TITLE
Get likes, shares, zaps for an event

### DISCUSSION THREAD

**Discussion by nima1981** [0xtrr/nostr-tool#50] [Q&A]
Is there any way using nostr-tool to retrieve the count of comments, likes, and shares, and the count and amount of zaps for any given event?

**Comment by 0xtrr** [comment]
Hey @nima1981!
 
It's not possible by using nostr-tool alone as we haven't added support for NIP-45 yet. However, we can do it with the help of `jq`.

### Reactions (likes)
```bash
nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 7 -o likes.json
cat likes.json | jq length
```

### Comments
```bash
nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 1 -o comments.json
cat comments.json | jq length
```

### Reposts (not quote reposts)
```bash
nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 6 -o reposts.json
cat reposts.json | jq length
```

### Zaps
```bash
nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 9735 -o zapreceipts.json
cat zapreceipts.json | jq length
```

It gets pretty complicated to sum up the total sats earned using the command-line so I don't have any good way of doing that with nostr-tool atm.

**Reply by nima1981** [reply]
Awesome, thanks! Does this only work with the nostr.oxtr.dev relay or can I use mulitple relays for this as well? What's jq? Thanks much for the quick reply.

**Reply by 0xtrr** [reply]
No you can use whatever relay you want. Just use the -r option before list-events and specify the relay you want to fetch the data from. There is support for fetching from multiple relays too by just adding another -r flag but it prints every single event it gets so you will get a lot of duplicates if the relays you are using has the same events in their database:

```bash
nostr-tool -r wss://nostr.oxtr.dev -r wss://relay.damus.io list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 9735 -o zapreceipts.json
```

`jq` is a command line tool to work with JSON documents. It's the swiss army knife of command-line JSON processing. You can read more about it here: https://jqlang.github.io/jq/

**Reply by nima1981** [reply]
Awesome, thanks much for the explanation!

**Reply by nima1981** [reply]
Hi, so I tried running this command:

`nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 7 -o likes.json`

It returns an error saying:

error: unexpected argument '-o' found

**Reply by 0xtrr** [reply]
Which version of nostr-tool do you use? You can check by running `nostr-tool -V`

**Comment by nima1981** [comment]
I'm using 0.3.0
What's the best way to upgrade? I think I installed it using these instructions:
cargo install nostr-tool
cargo build --release

Thanks!

**Reply by 0xtrr** [reply]
Ah, I believe the `-o` option is only available in 0.4.0 but I haven't released that yet (I'll try to do that later today or tomorrow). 
For now, you have to clone the repo and do `cargo build --release` in the root of the folder. Then you can run the program with `./target/release/nostr-tool -r wss://nostr.oxtr.dev list-events`

**Comment by nima1981** [comment]
Ah ok great, from where should I clone the repo? It seems like on here only up to 0.3.0 is available.

And after you release the new one is there a different way of accomplishing this that's easier?

Thanks!

**Reply by 0xtrr** [reply]
If you wait until I have released 0.4.0, you should be able to just do `cargo install nostr-tool` to update it to the latest version.

If you don't want to wait, you can do the following (assuming you're on a linux machine, idk how it works on windows):
1. `git clone https://github.com/0xtrr/nostr-tool.git`
2. `cd nostr-tool`
3. `cargo build --release`
4. `./target/release/nostr-tool [your commands and options here as usual)`

**Reply by nima1981** [reply]
OK great! After running ./target/release/nostr-tool -V it returns version 0.3.0 still so maybe it's best I'll wait until it's release on crates and then I try that option.

**Reply by thesimplekid** [reply]
It will still return 0.3.0 since the version isn't changed in the Cargo.toml. But if you've cloned and and built from main it's the newest.

**Comment by nima1981** [comment]
Well, just for funsies I tried this command  ./target/release/nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 7 -o likes.json

It doesn't return the -o unknown parameter error anymore, but it seems to just sit there hanging instead.

Also, what is the intended usage of the following parameters:
-e
-p
-d

**Reply by thesimplekid** [reply]
>It doesn't return the -o unknown parameter error anymore, but it seems to just sit there hanging instead.

That was happening for me earlier with oxtr.dev but it just worked for me now, maybe try again or a different relay. Usually if its hanging the relay isn't responding or responding slowly.   

> -e 

Is for querying event tags, any event that references the given event id is returned. `-p` is similar but for pubkeys. https://github.com/nostr-protocol/nips/blob/master/10.md

> -d 

Is used to query parametrized replaceable events tags.

https://github.com/nostr-protocol/nips/blob/master/33.md

**Reply by nima1981** [reply]
Great, thanks! Is there a way to find an event by note ID in this format: note1ecpkq3zkzcukeaw0q5t72puy96c365h9hfcu4x77227z08aqx5hq2d0g59

**Reply by thesimplekid** [reply]
Looks like it didnt handle the bech32 encoding. You can try my branch #53, or use 
`./target/release/nostr-tool  convert-key --key note1ecpkq3zkzcukeaw0q5t72puy96c365h9hfcu4x77227z08aqx5hq2d0g59 --to-hex` 
to convert it to the hex id `ce0360445616396cf5cf0517e507842eb11d52e5ba71ca9bde52bc279fa0352e` then run 
`./target/release/nostr-tool -r wss://relay.nostr.band list-events -e ce0360445616396cf5cf0517e507842eb11d52e5ba71ca9bde52bc279fa0352e`

**Reply by nima1981** [reply]
Awesome, thanks for that! One thing I noticed is that the json response seems to be an invalid json string, it looks like the one below, missing the quotation marks around the main level array indexes which caused my php script to fail to parse it as json and return null. A secondary issue is that there should be brackets around this whole statement as well, but that's an easy one to address on my end by just adding them manually.

Currently it looks like this:
 
```
0: {
  "id": "52faba302ba598bf7fa70e11490f2c7b18b1f6f4e58963a7fe12874213b116ca",
  "pubkey": "adc14fa3ad590856dd8b80815d367f7c1e6735ad00fd98a86d002fbe9fb535e1",
  "created_at": 1683078580,
  "kind": 7,
  "tags": [
    [
      "e",
      "3e269c300b94d042b85c21230b583c0713e66b71955753abfa9b02b586ddcb92"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "🤙",
  "sig": "4f6f228bb245684f7055e4fe4c30c338f003779fb396e341e5ab395c60322e0f295c2b0a672edc0648a54877cd5f0b214013774d59355e3bc62bb90391517dd4"
}
```

It should probably look like this:

```
{
"0": {
  "id": "52faba302ba598bf7fa70e11490f2c7b18b1f6f4e58963a7fe12874213b116ca",
  "pubkey": "adc14fa3ad590856dd8b80815d367f7c1e6735ad00fd98a86d002fbe9fb535e1",
  "created_at": 1683078580,
  "kind": 7,
  "tags": [
    [
      "e",
      "3e269c300b94d042b85c21230b583c0713e66b71955753abfa9b02b586ddcb92"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "🤙",
  "sig": "4f6f228bb245684f7055e4fe4c30c338f003779fb396e341e5ab395c60322e0f295c2b0a672edc0648a54877cd5f0b214013774d59355e3bc62bb90391517dd4"
}
}
```

**Reply by nima1981** [reply]
I'm able to resolve the invalid json by fixing the string via regex. But the one problem I keep running into is that for certain events the command just sits there and hangs forever and it stops the whole script. Would it be possible to incorporate some sort of timeout in this? Or maybe something else is wrong somewhere, it doesn't seem to make sense for it to be hanging forever in any scenario.

**Reply by nima1981** [reply]
One more issue I noticed in the formatting of the json string: if there's more than one array index, there's no comma before the subsequent array indexes. For example it just generated the following:

```
0: {
  "id": "ad2e020061946d7d18da1216209c8c01e71c20aba41168c01da683a74538d543",
  "pubkey": "6193172b8a7430d9d8c3f93ab5d85c2cb0451ce4c30ff1ea0a3849ae45b71a4c",
  "created_at": 1683115886,
  "kind": 7,
  "tags": [
    [
      "e",
      "8527b7dc3bd3b2fc1cb3ee443aac7e8b75d22237da3566f17ed81f94d2d97472"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "+",
  "sig": "ef33ce68e99fa77068aaff8a643eac6649ae424d5147e0b26f1cfffea22877be5775e19782e7c6ecf72b6e816a55b27e4a30be61112f10fe3c99cccc3642196f"
}
1: {
  "id": "4886079db3e4dabb897f0a616bdc0c326d16f9ef06d15631890ad962602e23e3",
  "pubkey": "0279278709f1becb38be09d9b6fe390ad816da5d72b0985a1a6e15003e3dd81d",
  "created_at": 1683116832,
  "kind": 7,
  "tags": [
    [
      "e",
      "8527b7dc3bd3b2fc1cb3ee443aac7e8b75d22237da3566f17ed81f94d2d97472"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "+",
  "sig": "64252507750275ba1155db8396e1e1c3658aafde8ecfd34c9541599d9492e84d2193bc37320fca613a35f8163b4f2331b34696b824220bd30013dccbc304d0c0"
}
```

Should probably be:

```
{
"0": {
  "id": "ad2e020061946d7d18da1216209c8c01e71c20aba41168c01da683a74538d543",
  "pubkey": "6193172b8a7430d9d8c3f93ab5d85c2cb0451ce4c30ff1ea0a3849ae45b71a4c",
  "created_at": 1683115886,
  "kind": 7,
  "tags": [
    [
      "e",
      "8527b7dc3bd3b2fc1cb3ee443aac7e8b75d22237da3566f17ed81f94d2d97472"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "+",
  "sig": "ef33ce68e99fa77068aaff8a643eac6649ae424d5147e0b26f1cfffea22877be5775e19782e7c6ecf72b6e816a55b27e4a30be61112f10fe3c99cccc3642196f"
},
"1": {
  "id": "4886079db3e4dabb897f0a616bdc0c326d16f9ef06d15631890ad962602e23e3",
  "pubkey": "0279278709f1becb38be09d9b6fe390ad816da5d72b0985a1a6e15003e3dd81d",
  "created_at": 1683116832,
  "kind": 7,
  "tags": [
    [
      "e",
      "8527b7dc3bd3b2fc1cb3ee443aac7e8b75d22237da3566f17ed81f94d2d97472"
    ],
    [
      "p",
      "d981591e0ea6153b8687b2aed670ab7d9b6c3ad018a360b2820b3cf0f7c0ae37"
    ]
  ],
  "content": "+",
  "sig": "64252507750275ba1155db8396e1e1c3658aafde8ecfd34c9541599d9492e84d2193bc37320fca613a35f8163b4f2331b34696b824220bd30013dccbc304d0c0"
}
}
```

**Reply by thesimplekid** [reply]
Ah I've always used this manually and not as part of a script so didn't catch some of these. I've updated #53 to include fixes for these it will print valid json now without the numbers just as a list of events. I've also added a `--timeout` argument so the timeout can be set in seconds.

**Reply by nima1981** [reply]
Awesome, thanks so much! What's #53? How do I get that version?

**Reply by nima1981** [reply]
I'm guessing I can just clone this one https://github.com/thesimplekid/nostr-tool/tree/list-events-bech32 right?

Then to deploy it I replace the previously downloaded folder with this one and build again?

Anything else I need to do, like delete/unlink the old one first?

**Reply by thesimplekid** [reply]
#53 is a pull request since I'm not the owner of this repo, @0xtrr  has to review and decide to merge into the repo or not. If you want it now anyway you can create a new branch `git checkout -b thesimplekid` and then pull the changes from my branch `git pull origin pull/53/head` and rebuild `cargo build -r` . To get back onto main if it gets merged `git checkout main` and rebuild

**Reply by thesimplekid** [reply]
>I'm guessing I can just clone this one https://github.com/thesimplekid/nostr-tool/tree/list-events-bech32 right?
 
you can do it this way to but think its generally better to not delete the main repo, and either do what I mentioned above or add my fork as a remote 

`git remote add thesimplekid https://github.com/thesimplekid/nostr-tool.git` 
`git fetch thesimplekid`
`git checkout list-events-bech32`

Just makes it easier to get back to main and not be stuck on a fork

**Reply by 0xtrr** [reply]
I have merged the PR now

**Reply by 0xtrr** [reply]
I have released version 0.4.0 now too, with the latest fix by @thesimplekid. You should be able to just do `cargo install nostr-tool` to get the latest version @nima1981.

**Reply by nima1981** [reply]
Done, works well, thanks very much!

**Reply by nima1981** [reply]
Question: Does it help in any way to submit the private key when running the commands with the list-events argument (like this one but adding a -p argument: nostr-tool -r wss://nostr.oxtr.dev list-events -e 8aa4836fd0efe87b7d14ba2e5709c7ed7120c0e3f24d4369aa5e2a8f305cd705 -k 7)

Or is that not needed at all when reading data from nostr?

**Reply by thesimplekid** [reply]
Keys are not needed for reading events from nostr relays for the most part. However, `nostr-sdk` which is the library this tool uses does need keys to create a client to read from relays. But what key you use does not matter and the same events will be returned so omitting the `-p` is fine and nostr-tool will generate a random key.  

The only exception to this is if a relay uses [NIP-42 Auth](https://github.com/nostr-protocol/nips/blob/master/42.md) in that case the private key would change the events that can be read. However, I don't believe nostr-tools supports that at the moment.

**Comment by nima1981** [comment]
Ah ok I understand now, I will give that a try. Thanks very much!!Sent from my iPhoneOn Jul 8, 2023, at 8:51 PM, thesimplekid ***@***.***> wrote:﻿
#53 is a pull request since I'm not the owner of this repo, @0xtrr  has to review and decide to merge into the repo or not. If you want it now anyway you can create a new branch git checkout -b thesimplekid and then pull the changes from my branch git pull origin pull/53/head and rebuild cargo build -r . To get back onto main if it gets merged git checkout main and rebuild
